File: FrontendAngular/src/app/services/mantenimiento.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

import { Mantenimiento } from '../models/mantenimiento';

@Injectable({
  providedIn: 'root'
})
export class MantenimientoService {

  private readonly url =
    'http://localhost:5234/api/mantenimiento';

  constructor(
    private http: HttpClient
  ) { }


  obtenerMantenimientos():
    Observable<Mantenimiento[]> {

    return this.http.get<Mantenimiento[]>(
      this.url
    );
  }


  obtenerMantenimiento(
    id: number
  ): Observable<Mantenimiento> {

    return this.http.get<Mantenimiento>(
      `${this.url}/${id}`
    );
  }


  obtenerPorVehiculo(
    idVehiculo: number
  ): Observable<Mantenimiento[]> {

    return this.http.get<Mantenimiento[]>(
      `${this.url}/vehiculo/${idVehiculo}`
    );
  }


  agregarMantenimiento(
    mantenimiento: Mantenimiento
  ): Observable<Mantenimiento> {

    const body = {
      idMantenimiento: 0,
      fecha: mantenimiento.fecha,
      tipo: mantenimiento.tipo,
      descripcion: mantenimiento.descripcion,
      costo: mantenimiento.costo,
      kilometraje: mantenimiento.kilometraje,
      estado: mantenimiento.estado,
      idVehiculo: mantenimiento.idVehiculo
    };

    return this.http.post<Mantenimiento>(
      this.url,
      body
    );
  }


  actualizarMantenimiento(
    mantenimiento: Mantenimiento
  ): Observable<Mantenimiento> {

    const body = {
      idMantenimiento:
        mantenimiento.idMantenimiento,

      fecha:
        mantenimiento.fecha,

      tipo:
        mantenimiento.tipo,

      descripcion:
        mantenimiento.descripcion,

      costo:
        mantenimiento.costo,

      kilometraje:
        mantenimiento.kilometraje,

      estado:
        mantenimiento.estado,

      idVehiculo:
        mantenimiento.idVehiculo
    };

    return this.http.put<Mantenimiento>(
      `${this.url}/${mantenimiento.idMantenimiento}`,
      body
    );
  }


  eliminarMantenimiento(
    id: number
  ): Observable<unknown> {

    return this.http.delete(
      `${this.url}/${id}`
    );
  }
}

File: VehiculosREST/VehiculosREST/Models/Mantenimiento.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace VehiculosREST.Models;

public class Mantenimiento
{
    [Key]
    public int IdMantenimiento { get; set; }

    [Required]
    public DateTime Fecha { get; set; }

    [Required]
    [MaxLength(100)]
    public string Tipo { get; set; } = string.Empty;

    [Required]
    [MaxLength(200)]
    public string Descripcion { get; set; } = string.Empty;

    [Column(TypeName = "decimal(10,2)")]
    public decimal Costo { get; set; }

    [Column(TypeName = "decimal(10,2)")]
    public decimal Kilometraje { get; set; }

    public bool Estado { get; set; }

    [Required]
    public int IdVehiculo { get; set; }

    public Vehiculo? Vehiculo { get; set; }
}